scripts/generate_fish_completion.py

```diff
@@ -0,0 +1,293 @@
+
+        Args:
+            bazel: String containing a path the a bazel binary to run.
+            output_user_root: String path to user root directory used for
+                running bazel commands.
+        """
+        self._bazel = bazel
+        self._output_user_root = output_user_root
+        self._startup_options = self._get_options_from_bazel(
+            ('help', 'startup_options'))
+        self._bazel_help_completion_text = self._get_bazel_output(
+            ('help', 'completion'))
+        self._param_types_by_subcommand = self._get_param_types()
+        self._subcommands = self._get_subcommands()
+
+    def write_completion(self, output_file):
+        """Writes a Fish completion script for Bazel to an output file.
+
+        Args:
+            output_file: File object opened in a writable mode.
+        """
+        output_file.write(_FISH_BAZEL_HEADER)
+        output_file.write('set {} {}\n'.format(
+            _FISH_BAZEL_COMMAND_LIST_VAR,
+            ' '.join(c.name for c in self._subcommands)))
+        output_file.write(_FISH_BAZEL_SEEN_SUBCOMMAND_DEF)
+        for opt in self._startup_options:
+            opt.write_completion(output_file)
+        for sub in self._subcommands:
+            sub.write_completion(output_file)
+
+    def _get_bazel_output(self, args):
+        return subprocess.check_output(
+            (self._bazel, '--output_user_root={}'.format(
+                self._output_user_root)) + tuple(args),
+            text=True)
+
+    def _get_options_from_bazel(self, bazel_args, **kwargs):
+        output = self._get_bazel_output(bazel_args)
+        return list(
+            Arg.generate_from_help(
+                r'^\s*--(\[no\])?(?P<name>\w+)\s+\((?P<desc>.*)\)$', output,
+                **kwargs))
+
+    def _get_param_types(self):
+        param_types = {}
+        for match in re.finditer(
+                r'^BAZEL_COMMAND_(?P<subcommand>.*)_ARGUMENT="(?P<type>.*)"$',
+                self._bazel_help_completion_text, re.MULTILINE):
+            sub = self._normalize_subcommand_name(match.group('subcommand'))
+            param_types[sub] = match.group('type')
+        return param_types
+
+    def _get_subcommands(self):
+        subs = []
+        output = self._get_bazel_output(('help', ))
+        block = re.search(r'Available commands:(.*\n\n)', output, re.DOTALL)
+        for sub in Arg.generate_from_help(
+                r'^\s*(?P<name>\S+)\s*(?P<desc>\S+.*\.)\s*$',
+                block.group(1),
+                is_subcommand=True):
+            sub.sub_opts = self._get_options_from_bazel(
+                ('help', sub.name), expected_subcommand=sub.name)
+            sub.sub_params = self._get_params(sub.name)
+            subs.append(sub)
+        return subs
+
+    _BAZEL_QUERY_BY_LABEL = {
+        'label': r'//...',
+        'label-bin': r'kind(".*_binary", //...)',
+        'label-test': r'tests(//...)',
+    }
+
+    def _get_params(self, subcommand):
```

Comment:
nit: That function is long enough to deserve a docstring.

Reply:
Done!